core/sr-primitives/src/lib.rs

```diff
@@ -636,53 +623,127 @@ impl From<ed25519::Signature> for AnySignature {
 	}
 }
 
-#[derive(Eq, PartialEq, Clone, Copy, Decode)]
+#[derive(Eq, PartialEq, Clone, Copy, Decode, Encode)]
 #[cfg_attr(feature = "std", derive(Debug, Serialize))]
-#[repr(u8)]
-/// Outcome of a valid extrinsic application. Capable of being sliced.
-pub enum ApplyOutcome {
-	/// Successful application (extrinsic reported no issue).
-	Success = 0,
-	/// Failed application (extrinsic was probably a no-op other than fees).
-	Fail = 1,
+/// Reason why an extrinsic couldn't be applied (i.e. invalid extrinsic).
+pub enum ApplyError {
+	/// General error to do with the permissions of the sender.
+	NoPermission,
+
+	/// General error to do with the state of the system in general.
```

Comment:
Can you elaborate on this? "Bad state" sounds a bit ambigious... 

For example:
1. is a corrupted storage value counts as example of `BadState`? 
1. What if an extrinsic tries to transfer balance from an account that doesn't have enough funds (although it has funds to cover fees), should it also be `BadState`?

Reply:
yeah that's the sort of thing.